pkg/ebpf/network_filter.go

```diff
@@ -0,0 +1,195 @@
+					}
+				}
+			}
+		}
+
+		if err == nil {
+			blacklist = append(blacklist, filter)
+		} else {
+			log.Error(err)
+		}
+	}
+	return blacklist
+}
+
+// parsePortAndProtocolFilter checks for valid port(s) and protocol filters
+// and returns a port/port range, protocol, and the validity of those values
+func parsePortAndProtocolFilter(v string) (uint64, uint64, ConnTypeFilter, error) {
+	lowerPort, upperPort := uint64(0), uint64(0)
+	v = strings.ToUpper(v)
+	connTypeFilter := ConnTypeFilter{TCP: true, UDP: true}
+
+	switch {
+	case strings.HasPrefix(v, "TCP"):
+		connTypeFilter.UDP = false
+		v = strings.TrimPrefix(v, "TCP")
+	case strings.HasPrefix(v, "UDP"):
+		connTypeFilter.TCP = false
+		v = strings.TrimPrefix(v, "UDP")
+	}
+
+	// The defined port is a wildcard
+	v = strings.TrimSpace(v)
+	if v == "*" {
+		return lowerPort, upperPort, connTypeFilter, nil
+	}
+
+	// The defined port is a range
+	if strings.ContainsRune(v, '-') {
+		portRange := strings.Split(v, "-")
+
+		// invalid configuration
+		if len(portRange) != 2 {
+			return lowerPort, upperPort, connTypeFilter, fmt.Errorf("invalid port range %v", portRange)
+		}
+		lowerPort, err := strconv.ParseUint(strings.TrimSpace(portRange[0]), 10, 16)
+		if err != nil {
+			return lowerPort, upperPort, connTypeFilter, fmt.Errorf("error parsing port: %s", err)
+		} else if lowerPort == 0 {
+			return lowerPort, upperPort, connTypeFilter, fmt.Errorf("invalid port %d", lowerPort)
+		}
+		upperPort, err := strconv.ParseUint(strings.TrimSpace(portRange[1]), 10, 16)
+		if err != nil {
+			return lowerPort, upperPort, connTypeFilter, fmt.Errorf("error parsing port: %s", err)
+		} else if upperPort == 0 {
+			return lowerPort, upperPort, connTypeFilter, fmt.Errorf("invalid port %d", upperPort)
+		}
+
+		// invalid configuration
+		if lowerPort > upperPort {
+			return lowerPort, upperPort, connTypeFilter, fmt.Errorf("invalid port range %d-%d", lowerPort, upperPort)
+		}
+
+		return lowerPort, upperPort, connTypeFilter, nil
+	}
+
+	// The defined port is an integer
+	lowerPort, err := strconv.ParseUint(v, 10, 16)
+	upperPort = lowerPort
+	if err != nil {
+		return lowerPort, upperPort, connTypeFilter, fmt.Errorf("error parsing port: %s", err)
+	} else if lowerPort == 0 {
+		return lowerPort, upperPort, connTypeFilter, fmt.Errorf("invalid port %d", lowerPort)
+	}
+	return lowerPort, upperPort, connTypeFilter, nil
+}
+
+// IsBlacklistedConnection returns true if a given connection should be excluded
+// by the tracer based on user defined filters
+func IsBlacklistedConnection(scf []*ConnectionFilter, dcf []*ConnectionFilter, conn *ConnectionStats) bool {
+	// No filters so short-circuit
+	if len(scf) == 0 && len(dcf) == 0 {
+		return false
+	}
+
+	if len(scf) > 0 && conn.Source != nil {
+		if findMatchingFilter(scf, util.NetIPFromAddress(conn.Source), conn.SPort, conn.Type) {
+			return true
+		}
+	} else if len(dcf) > 0 && conn.Dest != nil {
```

Comment:
should this be a separate if statement instead of an `else if` ? If the `len(scf) > 0` then destination filters will never be evaluated.

This is only passing tests because `conn.Source == nil` in the test cases, which never happens when this runs outside of tests.